internal/cody/subscription.go

```diff
@@ -0,0 +1,147 @@
+package cody
+
+import (
+	"context"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/featureflag"
+	"github.com/sourcegraph/sourcegraph/internal/ssc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+const USE_SSC_FEATURE_FLAG = "use-ssc-for-cody-subscription"
+
+type UserSubscriptionPlan string
+
+const (
+	UserSubscriptionPlanFree UserSubscriptionPlan = "free"
+	UserSubscriptionPlanPro  UserSubscriptionPlan = "pro"
+)
+
+type UserSubscription struct {
+	// Status is the current status of the subscription. "pending" means the user has no Cody Pro subscription.
+	// "pending" subscription will be removed post Feb 15, 2024. It is required to support users who have opted for
+	// Cody Pro Trial on dotcom.
+	Status ssc.SubscriptionStatus
+	// Plan is the plan the user is subscribed to. "free" or "pro".
+	Plan UserSubscriptionPlan
+	// ApplyProRateLimits is true if the user has a valid subscription status.
+	ApplyProRateLimits bool
+	// CurrentPeriodStartAt is the start date of the current billing cycle.
+	CurrentPeriodStartAt time.Time
+	// CurrentPeriodEndAt is the end date of the current billing cycle.
+	CurrentPeriodEndAt time.Time
+}
+
+// consolidateSubscriptionDetails consolidates the subscription details from dotcom and SSC.
+// Post Feb 15, 2024, this function will be removed and the logic will be simplified to us the SSC info only.
+func consolidateSubscriptionDetails(ctx context.Context, user types.User, subscription *ssc.Subscription) (*UserSubscription, error) {
+	now := currentTimeFromCtx(ctx)
+	febReleaseDate := time.Date(2024, 02, 14, 23, 59, 59, 59, now.Location())
+	isAfterFebReleaseDate := now.After(febReleaseDate)
+
+	// The user has put in their cc and signed up for Cody Pro on SSC
+	if subscription != nil {
+		currentPeriodStart, err := time.ParseInLocation(time.RFC3339, subscription.CurrentPeriodStart, now.Location())
```

Comment:
The ISO string sent across the wire will include the timezone already. But unless this is something we already do in the dotcom backend, I would strongly urge us to not localize that time to the user's location. And just pass it around as being in UTC. That way, it's clear what the frontend is receiving. (And the _browser_ is responsible for putting that into the user's locale.) ... and I'd be shocked if the dotcom server's clock is not using UTC for its timezone/location either.

```suggestion
		currentPeriodStart, err := time.ParseInLocation(time.RFC3339, subscription.CurrentPeriodStart, time.UTC)
```

Reply:
It mostly handles timezone craziness within a unit test in a CI environment.

Reply:
Updated.